[core-lro] Remove support for cancellation in LroEngine

deyaaeldeen opened the pull request:
cancelOperation is being deprecated in https://github.com/Azure/azure-sdk-for-js/pull/22753 so `LroEngine` no longer needs to have support for it. Note that this support was introduced in 2.3.0-beta.1 so removing it is not a breaking change.

Furthermore, this PR prepares for release today.

azure-sdk commented:
**API change check**

APIView has identified API level changes in this PR and created following API reviews.

[azure-core-lro](https://apiview.dev/Assemblies/Review/cfd7ec6ff277496ebadfb26ae3f31afe)

xirzec commented on sdk/core/core-lro/src/poller.ts:
```diff
@@ -379,13 +362,37 @@ export abstract class Poller<TState extends PollOperationState<TResult>, TResult
+  private processUpdatedState(): void {
+    if (this.operation.state.error) {
+      this.stopped = true;
+      this.reject!(this.operation.state.error);
```
any reason to have this be `this.reject!` instead of `this.reject?`

deyaaeldeen replied:
`this.reject` is initialized in the constructor (L255) so we know for sure it can't be `undefined`:
https://github.com/Azure/azure-sdk-for-js/blob/c59084407bf15d4f9c8d336eb3d73f36496a4766/sdk/core/core-lro/src/poller.ts#L247-L255

xirzec reviewed (commented):
I'm good with these changes, glad we could avoid the break, just a minor remark

deyaaeldeen commented on sdk/core/core-lro/src/lroEngine/operation.ts:
```diff
@@ -135,8 +134,7 @@ export class GenericPollOperation<TResult, TState extends PollOperationState<TRe
-    await this.cancelOp?.(this.state);
-    return this;
+    throw new Error("This method is deprecated because it wasn't implemented");
   }
```
It is a change in behavior for sure but is it necessarily a breaking change? Shouldn't we alert those customers that they're not actually canceling operation when they call `cancelOperation`?

joheredi replied:
Yeah, we definitely want to alert our customers, but I think throwing an error is too severe for this situation. It will turn a success path into a failure path with no code changes in consumers.

 I don't think the amount of disruption this error can cause pays off for letting the user know that this is basically a no-op. If we were throwing to fail fast (i.e .cancel would eventually cause an exception anyway) then it should be fine.

joheredi replied:
I think logging a warning should be a good enough alert for a minor version bump.

deyaaeldeen replied:
Ok, addressed in [d099a86](https://github.com/Azure/azure-sdk-for-js/pull/22844/commits/d099a869c47792376ce694f4ce4843a22a4c62a4)

joheredi reviewed (approved):
Thanks for sending out this change and addressing the feedback! Would you mind also marking cancel as deprecated in the `PollOperation` interface definition?

deyaaeldeen commented:
> Thanks for sending out this change and addressing the feedback! Would you mind also marking cancel as deprecated in the PollOperation interface definition?

I am doing the deprecation in https://github.com/Azure/azure-sdk-for-js/pull/22753 to be released later. I will deprecate `PollOperationState.cancel` there too.
